### ISSUE TITLE
DEVC "GAUGE HEAT FLUX GAS"

### ISSUE THREAD

**Issue by rupole1185** [firemodels/fds#13866]
**Describe the bug**
The following fds simulation has a two different behaviour: if the DEVC "GAUGE HEAT FLUX GAS" are present in the file the simulation is about 50% slower than the same simulation without these devices

**To Reproduce**
Run the analysis with and without the DEVC "GAUGE HEAT FLUX GAS" on FDS 6.9.1 and FDS nighly

**Expected behavior**
I wonder if such a slow down is normal or it could be reduced. What I notice is the CPU usage decreases with time in the simulation with DEVC, while it does not decreas with time without DEVC
![image](https://github.com/user-attachments/assets/cb1f88ce-a99e-478b-80f1-27a5724f915c)

**Desktop (please complete the following information):**
 - Version: FDS 6.9.1 and nightly

[TestGAUGE_TestGAUGE_Cleaned.fds.txt](https://github.com/user-attachments/files/18080716/TestGAUGE_TestGAUGE_Cleaned.fds.txt)

**Comment by Er9y714** [comment]
Did you compare the timestep and number of iterations?

**Comment by rupole1185** [comment]
In two example I have made I noticed that they got to 900 iterations (46.38s) in just two different time.
With the gauge it took it about 30 minutes, with without it it took approximately 15 minutes

**Comment by mcgratta** [comment]
I cannot run the case. There are vents that are not attached to any surface.
```
ERROR(820): VENT West Test Exhaust_01 has no solid backing or an orientation index (IOR) is needed. (MPI Process: 7, CHID: Test_GaugeDevice_with)
```

**Comment by rupole1185** [comment]
Hi Kevin,

yeah sorry, you have to remove the following 6 vents:

![image](https://github.com/user-attachments/assets/a60a1d15-97f3-4c98-b95b-6ad0fe315a0c)

Beside that, I have retested it with today's nightly and still the slow down when GAUGE devices are present is measurable. The simulations are about 20% slower with these devices.

**Comment by mcgratta** [comment]
OK, I'll run the cases. My first guess is that by using these "_GAS" devices, you are creating some overhead having to do with particles. Any "_GAS' device is a particle that records information.

**Comment by mcgratta** [comment]
By the way, you can see the timings of the major subroutines in the file `job_name_cpu.csv`

**Comment by mcgratta** [comment]
Could you post the `_cpu.csv` files for the cases that show a large difference in CPU time.

**Comment by rupole1185** [comment]
Hi there,

sorry but I have had busy days before Christmas. Well, I should re-generate
the CPU files ... I'll try to do that after Christmas!
Merry Christmas to everyone!

*Ruggero Poletto - **Managing director*

*VideoCall*: calendly.com/ruggero-poletto | *Mobile*: +39 378 30 33 133

________________________________________

CFD FEA SERVICE SRL P.IVA 04545570238 - via Borgo Grande 19, 37044 Cologna
Veneta VR

*PRIVACY and statement*

All information in this e-mail message, as well as any files included, are
intended for the exclusive use of the recipient (s) indicated above /
indicated, and therefore subject to protection pursuant to law no. 675 and
of the Legislative Decree of 30 June 2003 n. 196. Anyone who receives this
message by mistake, is requested to notify the writer of the incorrect
receipt, and proceed with the immediate cancellation of the same from the
system. Your cooperation is appreciated. Check our privacy policy here
<https://cloudhpc.cloud/privacy-policy-2/>.

**Comment by mcgratta** [comment]
Any progress on this issue?

**Comment by rupole1185** [comment]
Hi there,

Sorry but I am quite busy in this period and I could complete the cpu files only now.
I also have pending the cluster part but it is taking longer than expected as I have to update my system infrastructure to verify that and in the first attempt I tried it didn't work. I'll allocate some time to it as soon as I can.

[Test_GaugeDevice_cpu.csv](https://github.com/user-attachments/files/18504580/Test_GaugeDevice_cpu.csv)
[Test_NOGaugeDevice_cpu.csv](https://github.com/user-attachments/files/18504581/Test_NOGaugeDevice_cpu.csv)

**Comment by drjfloyd** [comment]
These files show the gauge case takes 12300 s and the no gauge case 12210 s. This is a difference of ~0.7 %.  This is consistent with Kevin's results of a difference of ~0.8 %.

**Comment by rupole1185** [comment]
Yeah,

but this is not consistent with the time step of my two analyses. I mean,
the one with the GAUGE calculated about 95s while the one without the GAUGE
reached about 238s in a similar amount of time (3.5hours).

WITH GAUGE:
[image: image.png]

WITHOUT GAUGE:
[image: image.png]

*Ruggero Poletto - **Managing director*

*VideoCall*: calendly.com/ruggero-poletto | *Mobile*: +39 378 30 33 133

________________________________________

CFD FEA SERVICE SRL P.IVA 04545570238 - via Borgo Grande 19, 37044 Cologna
Veneta VR

*PRIVACY and statement*

All information in this e-mail message, as well as any files included, are
intended for the exclusive use of the recipient (s) indicated above /
indicated, and therefore subject to protection pursuant to law no. 675 and
of the Legislative Decree of 30 June 2003 n. 196. Anyone who receives this
message by mistake, is requested to notify the writer of the incorrect
receipt, and proceed with the immediate cancellation of the same from the
system. Your cooperation is appreciated. Check our privacy policy here
<https://cloudhpc.cloud/privacy-policy-2/>.

**Comment by drjfloyd** [comment]
No images were attached. Can you attach the steps.csv files generated by these cases?  Those show the time FDS took for each time step.

**Comment by rupole1185** [comment]
Here you are!

*Ruggero Poletto - **Managing director*

*VideoCall*: calendly.com/ruggero-poletto | *Mobile*: +39 378 30 33 133

________________________________________

CFD FEA SERVICE SRL P.IVA 04545570238 - via Borgo Grande 19, 37044 Cologna
Veneta VR

*PRIVACY and statement*

All information in this e-mail message, as well as any files included, are
intended for the exclusive use of the recipient (s) indicated above /
indicated, and therefore subject to protection pursuant to law no. 675 and
of the Legislative Decree of 30 June 2003 n. 196. Anyone who receives this
message by mistake, is requested to notify the writer of the incorrect
receipt, and proceed with the immediate cancellation of the same from the
system. Your cooperation is appreciated. Check our privacy policy here
<https://cloudhpc.cloud/privacy-policy-2/>.

**Comment by drjfloyd** [comment]
The times in these step files are not consistent with the times in the cpu files. Are these from the same runs?

**Comment by mcgratta** [comment]
This is becoming confusing. 

@rupole1185 please post two input files. I will run both with the latest code and see if something looks unusual.

**Comment by rupole1185** [comment]
Here they are

[TestGAUGE_TestGAUGE_Cleaned.fds.txt](https://github.com/user-attachments/files/18511470/TestGAUGE_TestGAUGE_Cleaned.fds.txt)
[TestNOGAUGE_TestNOGAUGE_Cleaned.fds.txt](https://github.com/user-attachments/files/18511469/TestNOGAUGE_TestNOGAUGE_Cleaned.fds.txt)

**Comment by mcgratta** [comment]
I am running the cases now.

**Comment by mcgratta** [comment]
I have been running the cases since yesterday, and so far the end of the output (.err) files look like this:
```
==> Test_GaugeDevice.err <==
 Time Step:   28500, Simulation Time: 1468.6520 s

==> Test_NoGaugeDevice.err <==
 Time Step:   28600, Simulation Time: 1473.8052 s
```
which tells me that the "no gauge" case is running at a rate that is very slightly greater than the "gauge" case.

**Comment by rupole1185** [comment]
Not sure what to say. I have not tested the latest nightly, but one in
december and I had this difference.

*Ruggero Poletto - **Managing director*

*VideoCall*: calendly.com/ruggero-poletto | *Mobile*: +39 378 30 33 133

________________________________________

CFD FEA SERVICE SRL P.IVA 04545570238 - via Borgo Grande 19, 37044 Cologna
Veneta VR

*PRIVACY and statement*

All information in this e-mail message, as well as any files included, are
intended for the exclusive use of the recipient (s) indicated above /
indicated, and therefore subject to protection pursuant to law no. 675 and
of the Legislative Decree of 30 June 2003 n. 196. Anyone who receives this
message by mistake, is requested to notify the writer of the incorrect
receipt, and proceed with the immediate cancellation of the same from the
system. Your cooperation is appreciated. Check our privacy policy here
<https://cloudhpc.cloud/privacy-policy-2/>.

**Comment by rupole1185** [comment]
Hi there,

I have just checked with the Nighly version and attention: to run the analysis I sued the following number of cores and threads:

 _Number of MPI Processes:  8
 Number of OpenMP Threads: 2_

But I still have found a discrepancy in the advancement velocity of the analysis, with the simulation with GAUGE about 33% slower than the one without GAUGE.

![Image](https://github.com/user-attachments/assets/25ba5dbe-8fc4-4213-bf67-e9feb9a4b088)

[GAUGE.txt](https://github.com/user-attachments/files/18571096/GAUGE.txt)
[NOGAUGE.txt](https://github.com/user-attachments/files/18571097/NOGAUGE.txt)

**Comment by mcgratta** [comment]
I am running the cases now with 8 process and 2 OpenMP threads per process. I see only a slight increase in time for the GAUGE case.

![Image](https://github.com/user-attachments/assets/aa213247-50c0-46e6-8f2a-6a00adb5df93)

**Comment by rupole1185** [comment]
There must be something in the file writing procedure or in accessing the cache which is slowing down them on my machine then ..

**Comment by mcgratta** [comment]
Look at the output files `_cpu.csv`. Make a histogram of the time spent in the subroutines. Determine what the difference is between cases.

**Comment by rupole1185** [comment]
Comparing the two _cpu.csv files I get this:

![Image](https://github.com/user-attachments/assets/0a0f05e4-e131-4b01-bb19-fe79a3349b12)

The major difference comes from the COMM which is definitely higher than the version without the GAUGE

**Comment by mcgratta** [comment]
![Image](https://github.com/user-attachments/assets/d86d6dbf-b612-42bb-9b2f-1df4c368d5cd)

![Image](https://github.com/user-attachments/assets/44294c2d-8a16-4dfe-a8b3-0e1b652b0eac)

**Comment by rupole1185** [comment]
COMM I guess stands for communications, right? It means my analysis are spending time waiting for the meshes to communicate each other right?

**Comment by mcgratta** [comment]
Yes. I do not know why it is much worse with gauges, i.e. particles. What compilers are you using? OS? Fabric provider?

**Comment by drjfloyd** [comment]
Yes COMM is time spent in the routines that do MPI calls.  When you see some meshes with large COMM and some with small COMM that means those meshes with longer times finish their time step faster and spend more time in COMM waiting for the other meshes. The smallest COMM value would reflect actual time spent doing the communications. Your very large times suggest some issue with your system.

**Comment by rupole1185** [comment]
I do not compile FDS myself, I just use the precompiled versions I find on the nightly website ... I should try with a version I compile myself ...

**Comment by mcgratta** [comment]
Then it might be something to do with your network cards, Infiniband, etc. That is all black magic to me.

**Comment by mcgratta** [comment]
I'm closing the issue. If you have more information, reopen or start a new one.

**Comment by rupole1185** [comment]
Hi @mcgratta I investigated some more on this issue.

I have found potentially anouther couple of similar cases (it seems the issue is also present with the DEVC with RADIATIVE HEAT FLUX and VISIBILITY)

The interesting fact is that when I run the process on AMD processor I can notice the COMM time much higher than performing the same case on Intel processor.

I start thinking it depends on the processor infrastructure, but let me investigate more and I will give you more details.

**Comment by rupole1185** [comment]
Hi,

yeah if I remove the RADIATIVE HEAT FLUX device then I do not have the issue with the comm time. With the VISIBILITY device I suppose so, but I need more data to be sure. What is weird is that if I run everything on INTEL processors, everything works smoothly, also the initial issue with GAUGE HEAT FLUX. The issue appears just on AMD processors.

To share more files unluckily I need approvals from the users .. on Tuesday I have a video call with them and I'll see if they allow me.

**Comment by mcgratta** [comment]
Do you mean `RADIATIVE HEAT FLUX GAS` or `RADIATIVE HEAT FLUX`? There is a big difference.

**Comment by rupole1185** [comment]
Hi Kevin,

It's RADIATIVE HEAT FLUX. For the VISIBILITY I have a suspect, I am not
sure about it.

As I told you, so far I noticed this issue when running on AMD and not on
INTEL - where everything works perfectly actually.

*Ruggero Poletto - Managing director*

*VideoCall*: calendly.com/ruggero-poletto | *Mobile*: +39 378 30 33 133

________________________________________

CFD FEA SERVICE SRL UNIPERSONALE

P.IVA 04545570238 - via Borgo Grande 19, 37044 Cologna Veneta VR

*PRIVACY and statement*

All information in this e-mail message, as well as any files included, are
intended for the exclusive use of the recipient (s) indicated above /
indicated, and therefore subject to protection pursuant to law no. 675 and
of the Legislative Decree of 30 June 2003 n. 196. Anyone who receives this
message by mistake, is requested to notify the writer of the incorrect
receipt, and proceed with the immediate cancellation of the same from the
system. Your cooperation is appreciated. Check our privacy policy here
<https://cloudhpc.cloud/privacy-policy-2/>.

**Comment by mcgratta** [comment]
Can you post the input file where `RADIATIVE HEAT FLUX` is the only quantity that causes the run time to increase on the AMD processor.

**Comment by rupole1185** [comment]
I try to do it .. but it is a user input file and I need his approval to do
that. Tomorrow I have a video call with them and I see what I can do.

*Ruggero Poletto - Managing director*

*VideoCall*: calendly.com/ruggero-poletto | *Mobile*: +39 378 30 33 133

10/04 CODE ASTER USER MEETING | 15-16/04 FRENCH OPENFOAM
<https://www.foam-u.fr/> | 30/06-4/07 OPENFOAM WORKSHOP
<https://www.openfoam25.at/>

________________________________________

CFD FEA SERVICE SRL UNIPERSONALE

P.IVA 04545570238 - via Borgo Grande 19, 37044 Cologna Veneta VR

*PRIVACY and statement*

All information in this e-mail message, as well as any files included, are
intended for the exclusive use of the recipient (s) indicated above /
indicated, and therefore subject to protection pursuant to law no. 675 and
of the Legislative Decree of 30 June 2003 n. 196. Anyone who receives this
message by mistake, is requested to notify the writer of the incorrect
receipt, and proceed with the immediate cancellation of the same from the
system. Your cooperation is appreciated. Check our privacy policy here
<https://cloudhpc.cloud/privacy-policy-2/>.

**Comment by mcgratta** [comment]
Any news on this issue?

**Comment by rupole1185** [comment]
Hi Kevin,

no I do not have news in this regards and the for new case I had found I
was not allowed by the client to share it .. I'll monitor the situation and
keep you posted (it may take some time though)

*Ruggero Poletto - Managing director*

*VideoCall*: calendly.com/ruggero-poletto | *Mobile*: +39 378 30 33 133

10/04 CODE ASTER USER MEETING | 15-16/04 FRENCH OPENFOAM
<https://www.foam-u.fr/> | 19-20/06 PEW25
<https://iniped.com/home-2/events/pew25/> | 23-27/06 FDS Summer School |
30/06-4/07 OPENFOAM WORKSHOP <https://www.openfoam25.at/>

________________________________________

CFD FEA SERVICE SRL UNIPERSONALE

P.IVA 04545570238 - via Borgo Grande 19, 37044 Cologna Veneta VR

*PRIVACY and statement*

All information in this e-mail message, as well as any files included, are
intended for the exclusive use of the recipient (s) indicated above /
indicated, and therefore subject to protection pursuant to law no. 675 and
of the Legislative Decree of 30 June 2003 n. 196. Anyone who receives this
message by mistake, is requested to notify the writer of the incorrect
receipt, and proceed with the immediate cancellation of the same from the
system. Your cooperation is appreciated. Check our privacy policy here
<https://cloudhpc.cloud/privacy-policy-2/>.
